### PR TITLE
+tck #232 explain which tests are mendatory to be "compliant"

### LINKED ISSUE #232: Clarification on which TCK tests are mandatory
Hello again,

I wonder if someone could clarify what tests must be run in order to say "X implementation has passed the TCK?"

There seems some ambiguity in the README:

https://github.com/reactive-streams/reactive-streams-jvm/blob/master/tck/README.md

"The TCK is split up into 4 files JUnit 4 test classes which should be extended by implementers"

Taking the RFC meaning of "should" I assume this means the expectation is they are implemented unless there is a very good reason not to.

But later on in the README it states with respect to subscriber whitebox verification:

"We do not expect all implementations to make use of the plain SubscriberWhiteboxVerification, using the SubscriberBlackboxVerification instead."

Which appears to contradict the earlier statement that all tests should be implemented.

Can someone clarify exact which of the tests are mandatory and which are optional to implement in order to claim that an implementation passes the TCK?

Cheers!

### PR DESCRIPTION
**Pull request by ktoso**
Resolves #232

### CODE REVIEW

File: tck/README.md [hunk 1]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
 }
 ```
 
+## Which verifications must be implemented by an compliant implementation?
```
Comment (ktoso) [inline tck/README.md]: Open for better suggestions for the section title here, ideas?

Comment (viktorklang) [inline tck/README.md]: s/an/a

File: tck/README.md [hunk 2]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+In order to be considered an Reactive Streams compliant we require implementations to cover their
```
Comment (viktorklang) [inline tck/README.md]: This sentence is unclear. What does it mean? s/an/a

File: tck/README.md [hunk 3]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+Publishers and Subscribers with TCK verifications. Of course, if your library is only including
```
Comment (viktorklang) [inline tck/README.md]: "Of course," does not add anything

File: tck/README.md [hunk 4]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+Subscribers, you do not have to implement Publisher tests. The same applies to `IdentityProcessorVerification` -
+you do not need to implement it if your library does not contain Processors.
+
+In the case of Subscriber Verification are two styles of verifications to available: Blackbox or Whitebox.
+We *strongly* recommend testing your `Subscriber` with the `SubscriberWhiteboxVerification` as it is able to
```
Comment (viktorklang) [inline tck/README.md]: s/We/It's

File: tck/README.md [hunk 5]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+verify most of the specification. The `SubscriberBlackboxVerification` should only be used as a fallback,
+once you are sure implementing the whitebox version will not be possible for your implementation - if that happens
```
Comment (viktorklang) [inline tck/README.md]: How does one know it is not possible?

File: tck/README.md [hunk 6]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+feel free to open a ticket on the `reactive-streams/reactive-streams-jvm` project explaining what stopped you from
```
Comment (viktorklang) [inline tck/README.md]: This "link" is not helpful, right?

File: tck/README.md [hunk 7]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+implementing the whitebox verification.
+
+In summary: we require implementations to use Verifications for the parts of the specification that they implement,
```
Comment (viktorklang) [inline tck/README.md]: I think the use of "we" should be removed to make things sound less us/them.

File: tck/README.md [hunk 8]
```diff
@@ -479,6 +479,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
...
+and suggest using the whitebox verification over blackbox for the subscriber whenever possible.
+
 ## Upgrade story
 
 **TODO** - What is our story about updating the TCK? How do we make sure that implementations don't accidentally miss some change in the spec, if the TCK is unable to fail verify the new behavior? Comments are very welcome, discussion about this is under-way in [Issue #99 – TCK Upgrade Story](https://github.com/reactive-streams/reactive-streams/issues/99).
```
Comment (viktorklang) [inline tck/README.md]: This TODO should not exist in the 1.0.0.final so make it an Issue instead (if still relevant).

Comment (ktoso) [inline tck/README.md]: This is already addressed in: https://github.com/reactive-streams/reactive-streams-jvm/pull/246/files#diff-4ce068920e11452416705ccb8f46ec78L484

Comment (viktorklang) [inline tck/README.md]: Great, thanks!

File: tck/README.md (hunk 2 above)
Comment (ktoso) [inline tck/README.md]: I'm trying to answer the question @purplefox had here, paraphrasing it: "which tests do I need to implement for you guys to believe me my framework works with reactive streams?"

Do you think the "compliant" word is confusing or a good one here?

File: tck/README.md (hunk 3 above)
Comment (ktoso) [inline tck/README.md]: Right, thanks

File: tck/README.md (hunk 4 above)
Comment (ktoso) [inline tck/README.md]: +1

File: tck/README.md (hunk 6 above)
Comment (ktoso) [inline tck/README.md]: right, made it a real link :)

File: tck/README.md (hunk 2 above)
Comment (viktorklang) [inline tck/README.md]: Can it be made even simpler? Something like: In order to be a Reactive Streams implementation, aside from following the specification, one must also make sure that the implementations of the Publisher and Subscriber pass the TCK?

File: tck/README.md [hunk 9]
```diff
@@ -479,6 +487,22 @@ public class SkippingIdentityProcessorTest extends IdentityProcessorVerification
 }
 ```
 
+## Which verifications must be implemented by a compliant implementation?
+In order to be considered an Reactive Streams compliant require implementations to cover their
```
Comment (viktorklang) [inline tck/README.md]: This sentence does not make sense to me, is there some word missing?

**Comment by viktorklang** [comment]
:+1: from me after addressing the last comment.
I'll open an issue on updating the conversational style to be neutral instead of using "we" and "you" etc.

**Comment by viktorklang** [comment]
@reactive-streams/contributors Given that this only clarifies the TCK documentation and has already been available for review for 4 days, I'm going to merge this within 24h unless someone yells "no!". Thanks.

File: tck/README.md (hunk 9 above)
Comment (ktoso) [inline tck/README.md]: It was missing word (reactive streams compliant implementation), but that leads to repetition.
So I'll go with `In order to be considered Reactive Streams compliant, implementations must cover their...`

Comment (viktorklang) [inline tck/README.md]: :+1:

**Comment by viktorklang** [comment]
@reactive-streams/contributors Merging without objection

**Comment by viktorklang** [comment]
@ktoso this needs a rebase before it can merge cleanly. Would you mind?

**Comment by ktoso** [comment]
@viktorklang rebased and pushed

**Comment by viktorklang** [comment]
Thanks @ktoso, I'll merge once Travis approves

**Pull request merged**